### PR TITLE
Rajat develop

### PR DESCRIPTION
**Pull request by rjtkp**

### CODE REVIEW

**Comment by rjtkp** [comment]
I am writing the balance function

**Comment by rjtkp** [comment]
We have to have the root public and accessible from the main code

**Comment by project-tuva** [comment]
Hi Raj, I'm sorry but I disagree with the idea of promoting root to the public level. The balance function is going to be a public member of the class BST. Its auxiliary function are going to be either private or public members of the same BST class. Why do you need to have root public?

File: .gitignore [hunk 1]
```diff
@@ -2,4 +2,17 @@
 *.x
 *.o
 *.inp
-
+*.so
```
Comment (project-tuva) [inline .gitignore]: Please remove all the extensions we won't use.

File: Makefile [hunk 2]
```diff
@@ -11,18 +11,18 @@ default: $(EXE)
...
 
 clean:
-	rm -rf ./src/*.o *.o src/*~ include/*~ *~ html latex $(EXE)
+	rm -rf ./src/*.o *.o src/*~ include/*~ *~ html latex a.out $(EXE)
 
 doc: Doxygen/doxy.in
 	doxygen $^
 
-
+check:./bst.x
```
Comment (project-tuva) [inline Makefile]: Good idea to add some tests in the makefile!

File: example [hunk 3]
```diff
@@ -0,0 +1,20 @@
+1   sApXI
```
Comment (project-tuva) [inline example]: I was wondering if you could move this file into the examples folder and append the .inp extension to its name.

File: data_generator.cc [hunk 4]
```diff
@@ -0,0 +1,41 @@
+#include <iostream>
```
Comment (project-tuva) [inline data_generator.cc]: Could you please move this file to the examples folder?

File: include/bst.h [hunk 5]
```diff
@@ -67,14 +67,14 @@ class BST{
 
 
   bool check_eq_keys(const K& a, const K& b){
-    if (a==b) return true;
+    if ( fabs(a-b)< TOL ) return true; // Overloading will fail when overloaded function templated with double
```
Comment (project-tuva) [inline include/bst.h]: I need to discuss with you about this point. If you experienced that overloading fails, we need understand why. Comparing values with fabs only works with numerical types (or at least "recastable" to them).

File: bst_main.cc [hunk 6]
```diff
@@ -2,23 +2,28 @@
...
 // #include <bst_iterator.h>
 //#include <bst_all.h>
 #include <ap_error.h>
 
 
 int main(){
 
+  BST<int, int> bst {};
   std::istringstream iss("34 34");
+  if (bst.isBalanced( )) std::cout << "balanced" << '\n';
```
Comment (project-tuva) [inline bst_main.cc]: I'd define the is_balance function within the class and make it call by the public "void balance()" function. I this way the only function which the user may need to call from the main is the "void balance()" one, which manages inside its definition all the possible cases.

File: include/bst.h [hunk 7]
```diff
@@ -442,7 +486,8 @@ void BST<K,V>::populate_tree(){
   while (std::getline(std::cin, line)){
     std::stringstream ss(line);
     if (ss >> k >> v){
-      int check = BST::insert_node(k,v); // check error or throw exception. to be cmpleted
```
Comment (project-tuva) [inline include/bst.h]: Ok, I still have to manage the error handling in c++ std

File: include/bst.h [hunk 8]
```diff
@@ -363,7 +369,45 @@ int BST<K,V>::insert_nodes(std::istream& i_str ){
...
+    else{
+      tmp = tmp->right.get();
+      BST::find(tmp, k);
+    }
+    return 1;
+  }
+}
 
 template <typename K, typename V>
 int BST<K,V>::cmp_key(Node * tmp, const K& k, const V& v, Node * tmpUp){
```
Comment (project-tuva) [inline include/bst.h]: we also need to make this function return void

File: working_notes.txt [hunk 9]
```diff
@@ -1,75 +0,0 @@
-NOTES:
```
Comment (project-tuva) [inline working_notes.txt]: please leave this file, we may need as a scratch paper to write notes ecc. We won't deliver it in the final version

File: python/exam_postcard_list0.txt [hunk 10]
```diff
@@ -0,0 +1,19 @@
+date:2013-09-16; from:Dewey; to:Happy;
```
Comment (project-tuva) [inline python/exam_postcard_list0.txt]: Please remove the python folder, we'll create another folder for it

**Comment by project-tuva** [comment]
I noticed that the requested printing layout is "key: value". I'll fix my code asap, cause I'm only printing the values.

File: include/bst.h (hunk 5 above)
Comment (project-tuva) [reply include/bst.h]: Fixed by leaving function declaration inside the class and moving function definitions outside it.
https://stackoverflow.com/questions/10863444/overriding-a-function-only-for-certain-types-in-template

**Comment by project-tuva** [comment]
I've written copy /move ctors and copy/move assignments.

**Pull request merged**